pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -327,14 +328,31 @@ func (configgen *ConfigGeneratorImpl) createGatewayHTTPFilterChainOpts(
 		// We know that this is a HTTPS server because this function is called only for ports of type HTTP/HTTPS
 		// where HTTPS server's TLS mode is not passthrough and not nil
 		for _, server := range servers {
+			// Workaround for race condition in Envoy where two listeners share same RDS object (l1, l2, rA).
+			// When one of them (l1) is being updated/drained and an RDS update (rA) arrives at same time,
+			// a race condition causes RDS update to refer to the no-longer-existing listener (l1-drained) -
+			// which in turn results in a segfault.
+			// This could happen when someone renames a server's port causing listener update (e.g., https-foo to https-bar),
+			// and renames another server's port to be the same as the previous port name (https->https-foo). If l1 was using server1
+			// with rds https.443.https-foo and l2 was using rds https.443.https, this renaming would result in
+			// new LDS push to the listener l1 , with same RDS resource name as the one referred to by the first listener that is being
+			// drained/updated.
```

Comment:
After I wrote this, I am not sure if this fix will help. Because we use filter chains in the gateway for a single HTTPS listener. We have different RDS for each filter chain, but they are all part of same listener.